### DISCUSSION TITLE
Toward a better document.currentScript

### DISCUSSION THREAD

**Discussion by WebReflection** [pyscript/pyscript#1930] [Proposals]
While working on [this issue](https://github.com/pyscript/pyscript/issues/1923) I've noticed various problems regarding the current magic mokey-patching of the `document.currentScript` accessor:

  * 3rd party scripts that might check, need, or use, such reference might be misleading. As example, if any *js_module* side effect with the presence of such reference, which is `null` in JS modules and in other circumstances, we can't really enforce that code to think the current python code is the one running the script
  * there's a natural caveat around `document.currentScript`: it must be referenced once ASAP because if accessed via listeners or lazily, it might not return the expected node, if any at all. This is a footgun already for the Web, I think we can do better than propagate such footgun to PyScript users too
  * as we're patching a native DOM API, it's very possible we'll eventually encounter changes or browsers that won't allow us to change that behavior ... it's a long-term slippery slope ...

## The Proposal

In [polyscript](https://pyscript.github.io/polyscript/#the-polyscript-module) the issue has been solved by attaching the known reference to the current script to the module once. This works because in there every script is a different interpreter, even if it's the same interpreter each time, unless the *env* is explicitly different **and**, in such case, first comes first serves as `currentScript` reference.

In **PyScript** though, all scripts share, by design, the same interpreter, hence *env*, at least on the main thread. This is the reason the `document.currentScript` is monkey-patched right before code gets executed, and cleared right after that execution.

### Why cannot workers do the same?

Long story short is that workers cannot monkey-patch on the main thread a synchronous accessor ... or better, the moment the main JS code is reached from workers it can be sync or async, it doesn't matter, but if a patch lands on the main thread, its returned value would inevitably be async, because the contract here is that `sync` works one way, but not the other.

As practical result, if the worker monkey patches the `document.currentScript` accessor on the main, as that's the `document` reference we all use and deal with in workers, the returned value would be a promise ... the main thread would need to resolve through the worker what is the current script that should be returned through the *getter* that passes through the worker to understand what that is, even if lives on main.

As it's hard to not have headaches just reading my last sentence, we want to do better ... so here are a few options:

## Proposal

  * `pyscript.get_current_script()` making it explicit this is a side-effects free operation (**edit** please see [the explanation around this point](https://github.com/pyscript/pyscript/discussions/1930#discussioncomment-8094249))
  * `pyscript.__script__` which doesn't necessarily make it explicit it's side effect free, because `from pyscript import __script__` might indeed return a different reference later on
  * `pyscript.current_script` which suffers all the same gotchas the previous point has ...

I think we should scratch out of the equation ambiguous solutions, but I haven't heard in today community call meeting other options, although we all seem to agree we don't need to make our users life harder, we should provide an utility that is granted to always return what's expected.

Well, this is not an easy task because code that runs in listeners might happen without giving us a chance to wrap the right current script around, but at least `from pyscript import get_current_script` and `pyscript.get_current_script` always return the same intent: a function that returns the current script.

### To think about ...

  * I am personally not sure the story around returning always the same reference is a super strong argument simply because if we return always the same reference but its execution `...()` might return something different, we're moving the problem instead of fixing it
  * whenever having `current_script` is needed or desired, the "*trap it ASAP*" solution works and can disambiguate with listeners too, if that reference is reached later on
  * the actual DOM standard around this topic is already full of caveats because of the very same reasons we have this issue ... we can't prevent bad code in user defined listeners, we could at least document why that `__script__` or `current_script` thing *MUST* be trapped ASAP instead ... still leading to less surprises compared to the globally mokey patched `document.currentScript`

- - -

If you have thoughts, ideas, or solutions, to share, you are more than welcome, but this discussion is kinda needed otherwise the resolution for the related issue is kinda blocked too.

Thank You!

/cc @antocuni @ntoll @fpliger @leriomaggio @JeffersGlass @madhur-tandon

**Comment by WebReflection** [comment]
Quick follow up ... **another possibility**: we just provide a `__script__` so that nobody needs to import it. It could be named `__current_script__` or anything, and it will be available within the executing code.

This would solve at least these caveats:

  * there is nothing to import ... it's just a reference
  * it can be consistent in both workers and cases where the main has only one PyScript element, which is the common use case

The documentation can then explain that "*first come first serve*" ... meaning that if the currently running code does not have already a reference to a `__script__` special reference, it gets created once before code gets executed.

That is basically:

```python
# top most pre-pended code
from polyscript import currentScript as __script__
```

That should be it:

  * in polyscript that reference is static per env ... meaning in PyScript is first-come first serve
  * if users really care about disambiguating the script because they have multiple scripts on the page, they can always use IDs or any other selector to be sure they reach the script

I can't think of anything better because we cannot literally intercept listeners or stuff that happens after other scripts around got executed and we don't have local context per each code evaluation within the same runtime so until that gets fixed, I believe there is no a "*fix them all*" solution but this one looks like the easiest to reason about.

I also forgot to cc @mchilvers in the previous discussion ... what do you all think about this latter approach? 🤔

**Comment by ntoll** [comment]
FWIW I'm +1 on `get_current_script()` +/-0 on `__script__`

**Reply by WebReflection** [reply]
a function doesn't work and it actually helps creating more confusion and issues because you feel like you can invoke it at any time and that's the issue we are trying to fix: you **cannot** invoke that function at any time, you **must** retrieve the `currentScript` ASAP otherwise that can be done by DOM standards.

This, unless we return the `polyscript.currentScript` which also has the issue that multiple scripts (pyodide or MicroPython) would get only the first one that run that code on *main*, in the *worker* we have never an issue.

**Reply by ntoll** [reply]
Hmm... OK now I'm confused (as usual). ;-)

You say:

> pyscript.get_current_script() making it explicit this is a side-effects free operation

Which sounds like it's exactly what we need.

**Reply by WebReflection** [reply]
in this regard, `__script__` is meant to trap ASAP and it's already available and immutable *but* if multiple scripts on main are meant then users must be aware that `__script__` points at the first one that bootstrapped the interpreter, whatever that is.

I hope this is clear: a function is more a footgun than a solution + we can't overload all runtime invocations or asynchronous things with the current script ... if we had context per script then maybe we could use a function *but* AFAIK we don't have that and that might also cause other issues, like not globally shared references anymore.

We can't have all things solved here, unfortunately.

**Reply by WebReflection** [reply]
> Which sounds like it's exactly what we need.

the side effects are related to the *trap ASAP* vs lazy import ... if the trap can change in the future that causes confusion.

```python
from pyscript import current_script
# this is working top most the code

def handler(event):
  from pyscript import current_script
  # this might already be different
  # when this gets executed

def another_one():
  import pyscript
  pyscript.current_script
  # this can have again another value
```

That's the side-effects I was referring to, apologies if not super clear.
A function avoids that confusion but its returned value is **inevitably inconsistent** accordingly with *when* you execute that function.

If script 1 set a click listener then script 2 gets executed that is the current script. When listener in script 1 triggers and you invoke that function you'll get the script 2 instead.


```python
from pyscript import get_current_script

get_current_script()
# this is working top most the code

def handler(event):
  get_current_script()
  # this might already return something else
  # when this gets executed

def another_one():
  import pyscript
  pyscript.get_current_script()
  # same callback, but the returned value is not same
  # unless it always returns the equivalent of __script__
```

There is no way around this. With an explicit `__script__` trapped ASAP it's always the same reference, never a different one.

**Reply by ntoll** [reply]
Aha. Understood. Then `__script__` is clearly the way to go.

**Reply by WebReflection** [reply]
@antocuni @JeffersGlass what are your thoughts around this further explanation with examples?

**Reply by WebReflection** [reply]
FWIW we could also play with the name to be even more explicit and in that regard a callback might be better or less *magic* (all names are as example, I don't have strong opinion on those names as long as meaningful enough):

  * `pyscript.get_initiator()` - a callback that returns the first element that bootstrapped the interpreter, either `<script type=x>` or `<x-script>` ... no semantic involved, it's the first node (on main) that bootstrapped the interpreter, it's the worker related element otherwise
  * `__initiator__` to mimic the same without import needed ... as the callback would always return the same node, this might be also OK (imho)
  * alternatively, we can also suggest to use `from polyscript import currentScript` once latest gets merged ... we delay this resolution/conversation, I improve the *polyscript* related documentation, we mention this somehow in our PyScript docs too
  * any other idea is welcome as long as the issue is clear and the solution easy at least for our users to understand

**Comment by WebReflection** [comment]
To whom it might concern, I am closing this discussion as there is currently no way to provide a reliable `currentScript` on the main thread that is meaningful for all the scripts sharing the same runtime, as these can be multiple, their callbacks or listeners can trigger at any point in time, and no matter if we use a function or `__script__`, that reference will always be something different on main until we manage to have some sort of block scope that also should avoid conflicts or issues with other expectations, as example the fact a variable defined in script No.1 is available in script No.2

This MR solves the main use case where `currentScript` was desired to reach the attached terminal reference and it's a much cleaner and robust approach: https://github.com/pyscript/pyscript/pull/1947
